### PR TITLE
Fix gh-1369; handling assignments in OpAsQuery

### PR DESCRIPTION
**Pull request by Aklakan**
This PR improves handling of assignents w.r.t. projections by taking the order into account.
Assignments will only be converted into projections if the projection references them in the same order.

There is one class cast to ElementGroup which I would like to get rid of and I would need input on what might be the cleanest way to do so.
```
// The following cast may be critical:
// The problem is that here its too late to add to currentGroup()
ElementGroup activeGroup = ((ElementGroup)query.getQueryPattern());
```

It adds two more test cases testBind08 and testBind09 that would fail with the prior implementation.
All other tests continue to work.

```
test_roundTripQuery("SELECT (?x AS ?y) { BIND ('x' AS ?x) }");
test_roundTripQuery("SELECT ('y' AS ?y) ?x { BIND ('x' AS ?x) }"); } // Note that ?x is assigned before ?y, but projected only afterwards
```

### CODE REVIEW

**Comment by Aklakan** [comment]
I added a few more tests for somewhat corner cases that came to my mind today; it's still green - so from my side its ready for review/feedback.

**Review by afs — APPROVED** [review]
One comment about a test to add but otherwise looks ready.

File: jena-arq/src/test/java/org/apache/jena/sparql/algebra/TestOpAsQuery.java [hunk 1]
```diff
@@ -42,206 +42,213 @@ public class TestOpAsQuery {
...
-    
-    @Test public void testOptional01() 
+
+    // https://github.com/apache/jena/issues/1369
+    @Test public void testBind08() { test_roundTripQuery("SELECT (?x AS ?y) { BIND ('x' AS ?x) }"); }
+    @Test public void testBind09() { test_roundTripQuery("SELECT ('y' AS ?y) ?x { BIND ('x' AS ?x) }"); }
+    @Test public void testBind10() { test_roundTripQuery("SELECT ('y' AS ?y) ?s ('x' AS ?x) { ?s ?p ?o }"); }
+    @Test public void testBind11() { test_roundTripQuery("SELECT ?s ('y' AS ?y) ?p ?x ?o { ?s ?p ?o BIND ('x' AS ?x) }"); }
+    @Test public void testBind12() { test_roundTripQuery("SELECT ?w ('y' AS ?y) ?x { BIND('w' AS ?w) ?s ?p ?o BIND ('x' AS ?x) }"); }
+
```
[the remark this answers is not in the document]
Comment (Aklakan) [reply jena-arq/src/test/java/org/apache/jena/sparql/algebra/TestOpAsQuery.java:75]: Done

**Comment by Aklakan** [comment]
I added the test case.
For robustness I replaced the naive class cast with a check to ensure that there exists an ElementGroup to which to append the ElementBinds.

```java
                    if (activeElement instanceof ElementGroup) {
                        activeGroup = (ElementGroup)activeElement;
                    } else {
                        // Not sure whether it's possible here for BINDs to exist with the
                        // activeElement NOT being a group pattern - but better safe than sorry
                        activeGroup = new ElementGroup();
                        activeGroup.addElement(activeElement);
                        query.setQueryPattern(activeGroup);
                    }
```
